pkg/diagnostics/networkpod/service.go

```diff
@@ -0,0 +1,162 @@
+		netnsList, err := d.OSClient.NetNamespaces().List(kapi.ListOptions{})
+		if err != nil {
+			r.Error("DSvcNet1006", err, fmt.Sprintf("Getting all network namespaces failed. Error: %s", err))
+			return r
+		}
+
+		vnidMap = map[string]uint32{}
+		for _, netns := range netnsList.Items {
+			vnidMap[netns.NetName] = netns.NetID
+		}
+	}
+
+	localGlobalPods, localNonGlobalPods := getGlobalAndNonGlobalPods(localPods, vnidMap)
+
+	// Applicable to flat and multitenant networks
+	if len(localGlobalPods) > 0 {
+		checkPodToServiceConnection(localGlobalPods[0], services[0], vnidMap, r)
+	} else {
+		r.Warn("DSvcNet1007", nil, fmt.Sprintf("Skipping service connectivity test for global projects. Reason: Couldn't find a global pod."))
+	}
+
+	// Applicable to multitenant network
+	isMultitenant := (vnidMap != nil)
+	if isMultitenant {
+		if len(localNonGlobalPods) > 0 {
+			checkPodToServiceConnection(localNonGlobalPods[0], services[0], vnidMap, r)
+		} else {
+			r.Warn("DSvcNet1008", nil, fmt.Sprintf("Skipping service connectivity test for non-global projects. Reason: Couldn't find a non-global pod."))
+		}
+	}
+	return r
+}
+
+func getAllServices(kubeClient *kclient.Client) ([]kapi.Service, error) {
+	filtered_srvs := []kapi.Service{}
+	serviceList, err := kubeClient.Services(kapi.NamespaceAll).List(kapi.ListOptions{})
+	if err != nil {
+		return filtered_srvs, err
+	}
+
+	for _, srv := range serviceList.Items {
+		if len(srv.Spec.ClusterIP) == 0 || srv.Spec.ClusterIP == kapi.ClusterIPNone {
+			continue
+		}
+		filtered_srvs = append(filtered_srvs, srv)
+	}
+	return filtered_srvs, nil
+}
+
+func printService(svc kapi.Service) string {
+	return fmt.Sprintf("%s/%s", svc.Namespace, svc.Name)
+}
+
+func checkPodToServiceConnection(fromPod kapi.Pod, toService kapi.Service, vmap map[string]uint32, r types.DiagnosticResult) {
+	if len(fromPod.Status.ContainerStatuses) == 0 {
+		r.Error("DSvcNet1009", nil, fmt.Sprintf("ContainerID not found for pod %q", printPod(fromPod)))
+		return
+	}
+
+	success := expConnStatus(fromPod.Namespace, toService.Namespace, vmap)
+
+	kexecer := kexec.New()
+	containerID := kcontainer.ParseContainerID(fromPod.Status.ContainerStatuses[0].ContainerID).ID
+	pid, err := kexecer.Command("docker", "inspect", "-f", "{{.State.Pid}}", containerID).CombinedOutput()
+	if err != nil {
+		r.Error("DSvcNet1010", err, fmt.Sprintf("Fetching pid for pod %q failed. Error: %s", printPod(fromPod), err))
+		return
+	}
+
+	// In bash, redirecting to /dev/tcp/HOST/PORT or /dev/udp/HOST/PORT opens a connection
+	// to that HOST:PORT. Use this to test connectivity to the service; we can't use ping
+	// like in the pod connectivity check because only connections to the correct port
+	// get redirected by the iptables rules.
+	srvConCmd := fmt.Sprintf("echo -n '' > /dev/%s/%s/%d", strings.ToLower(string(toService.Spec.Ports[0].Protocol)), toService.Spec.ClusterIP, toService.Spec.Ports[0].Port)
```

Comment:
oh, better yet, you can just use `net.Dial` to try connecting rather than calling a helper program